File: Chatsito/chatsito.Web/ToolDefinitionProvider.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace chatsito.Web
{
    public static class ToolDefinitionProvider
    {
        public static List<object> GetTools()
        {
            return new List<object>
            {
                new
                {
                    type = "function",
                    function = new
                    {
                        name = "search_web",
                        description = "Performs a web search for a given query. Returns a summary of relevant information along with URL citations. Use this when you need up-to-date information, news, or factual answers that you don't know.",
                        parameters = new
                        {
                            type = "object",
                            properties = new
                            {
                                query = new
                                {
                                    type = "string",
                                    description = "The search query to look up on the internet."
                                }
                            },
                            required = new[] { "query" }
                        }
                    }
                },
                new
                {
                    type = "function",
                    function = new
                    {
                        name = "get_current_date_time",
                        description = "Gets the current date and time.",
                        parameters = new
                        {
                            type = "object",
                            properties = new { },
                            required = Array.Empty<string>()
                        }
                    }
                },
                new
                {
                    type = "function",
                    function = new
                    {
                        name = "execute_http_request",
                        description = "Executes an arbitrary HTTP request (GET or POST) and returns the raw response body. Use this to fetch specific URLs, interact with APIs, or follow up on search results. You must parse the raw result yourself."
                        + " NOTE: please use sparringly. Unlike the search_web tool, this tool causes the context window to explode and provides a suboptimal user experience. Only execute this tool if it is strictly necessary or if the user explicitly asks you to look for something online.",
                        parameters = new
                        {
                            type = "object",
                            properties = new
                            {
                                url = new { type = "string", description = "The full URL to request." },
                                method = new { type = "string", description = "The HTTP method (GET or POST)." },
                                body = new { type = "string", description = "The request body for POST requests (optional)." }
                            },
                            required = new[] { "url", "method" }
                        }
                    }
                }
            };
        }

        public static async Task<string> ExecuteToolAsync(string name, IDictionary<string, object>? arguments)
        {
            if (name == "search_web")
            {
                string query = "";
                if (arguments != null && arguments.TryGetValue("query", out var queryObj) && queryObj != null)
                {
                    query = queryObj.ToString() ?? "";
                }
                return await ExecuteWebSearchAsync(query);
            }
            else if (name == "get_current_date_time")
            {
                return DateTime.Now.ToString("f");
            }
            else if (name == "execute_http_request")
            {
                string url = "";
                string method = "GET";
                string body = "";


                if (arguments != null)
                {
                    if (arguments.TryGetValue("url", out var urlObj) && urlObj != null) url = urlObj.ToString() ?? "";
                    if (arguments.TryGetValue("method", out var methodObj) && methodObj != null) method = methodObj.ToString() ?? "GET";
                    if (arguments.TryGetValue("body", out var bodyObj) && bodyObj != null) body = bodyObj.ToString() ?? "";
                }


                return await ExecuteHttpRequestAsync(url, method, body);
            }

            return $"Error: Tool '{name}' not found.";
        }

        private static async Task<string> ExecuteHttpRequestAsync(string url, string method, string body)
        {
            try
            {
                using var client = new HttpClient();
                client.DefaultRequestHeaders.Add("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36");


                HttpResponseMessage response;
                if (method.Equals("POST", StringComparison.OrdinalIgnoreCase))
                {
                    var content = new StringContent(body);
                    response = await client.PostAsync(url, content);
                }
                else
                {
                    response = await client.GetAsync(url);
                }


                if (!response.IsSuccessStatusCode)
                {
                    return $"HTTP Error: {(int)response.StatusCode} {response.ReasonPhrase}\n" + await response.Content.ReadAsStringAsync();
                }


                return await response.Content.ReadAsStringAsync();
            }
            catch (Exception ex)
            {
                return $"Error executing HTTP request: {ex.Message}";
            }
        }

        private static async Task<string> ExecuteWebSearchAsync(string query)
        {
            try
            {
                using var client = new HttpClient();
                client.DefaultRequestHeaders.Add("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36");
                var content = new FormUrlEncodedContent(new[] { new KeyValuePair<string, string>("q", query) });
                var response = await client.PostAsync("https://lite.duckduckgo.com/lite/", content);
                response.EnsureSuccessStatusCode();
                var html = await response.Content.ReadAsStringAsync();

                var results = new List<string>();


                var titleRegex = new System.Text.RegularExpressions.Regex(@"<a rel=""nofollow"" href=""([^""]+)"" class='result-link'>([^<]+)</a>");
                var snippetRegex = new System.Text.RegularExpressions.Regex(@"<td class='result-snippet'>([\s\S]*?)</td>");


                var titles = titleRegex.Matches(html);
                var snippets = snippetRegex.Matches(html);

                for (int i = 0; i < Math.Min(3, Math.Min(snippets.Count, titles.Count)); i++)
                {
                    string url = titles[i].Groups[1].Value;
                    string title = titles[i].Groups[2].Value.Trim();
                    string snippet = System.Text.RegularExpressions.Regex.Replace(snippets[i].Groups[1].Value, "<[^>]+>", "").Trim();
                    results.Add($"{i + 1}. {title}\nURL: {url}\nSnippet: {snippet}");
                }

                if (results.Count == 0)
                {
                    var ahrefs = System.Text.RegularExpressions.Regex.Matches(html, @"<a class=""result-snippet""[^>]*href=""([^""]+)""[^>]*>(.*?)</a>", System.Text.RegularExpressions.RegexOptions.Singleline);
                    int count = 0;
                    foreach (System.Text.RegularExpressions.Match m in ahrefs)
                    {
                        if (count >= 3) break;
                        string url = m.Groups[1].Value;
                        string title = System.Text.RegularExpressions.Regex.Replace(m.Groups[2].Value, "<[^>]+>", "").Trim();
                        results.Add($"{++count}. {title}\nURL: {url}");
                    }
                }

                if (results.Count == 0)
                {
                    return "No results found or unable to parse search engine results.";
                }

                return string.Join("\n\n", results);
            }
            catch (Exception ex)
            {
                return $"Error executing web search: {ex.Message}";
            }
        }
    }
}

File: Fixito/ToolDefinitionProvider.js
class ToolDefinitionProvider {
    constructor() {
        this.tools = [
            {
                name: 'searchWorkspace',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/searchWorkspace',
                adapter: async (args, impl) => await impl(args.query, args.fileGlob),
                getProgressText: args => 'Searching workspace for "' + (args.query || '') + '"...',
                schema: {
                    name: "searchWorkspace",
                    description: "Searches the current workspace files recursively for a specific text, code query, or symbol name. Returns matching lines and file paths. NOTE: Search is literal by default. Filter glob paths must use forward slashes.",
                    parameters: {
                        type: "object",
                        properties: {
                            query: {
                                type: "string",
                                description: "The literal text, keyword, or code symbol to search for in workspace files."
                            },
                            fileGlob: {
                                type: "string",
                                description: "Optional glob pattern to restrict target files, e.g. '**/*.cs' or '**/*.js'. Relative to the workspace root."
                            }
                        },
                        required: ["query"]
                    }
                }
            },
            {
                name: 'getFullGraph',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/getFullGraph',
                adapter: async (args, impl) => await impl(args.filePath, parseInt(args.lineNumber, 10), parseInt(args.characterNumber || '1', 10)),
                getProgressText: args => 'Generating full dependency call graph...',
                schema: {
                    name: "getFullGraph",
                    description: "Generates a full backward invocation dependency tree (call graph) for a symbol at a specific location, listing all callers recursively up to the root caller methods. Detects and stops on circular dependencies.",
                    parameters: {
                        type: "object",
                        properties: {
                            filePath: {
                                type: "string",
                                description: "The relative or absolute file path containing the symbol. Must point to a file that exists."
                            },
                            lineNumber: {
                                type: "integer",
                                description: "The 1-based line number of the symbol declaration or reference (1 represents the first line). Must be positive."
                            },
                            characterNumber: {
                                type: "integer",
                                description: "Optional. The 1-based character/column offset of the symbol on that line (1 represents the first character). Must be positive. Defaults to 1."
                            }
                        },
                        required: ["filePath", "lineNumber"]
                    }
                }
            },
            {
                name: 'getCurrentEditorContext',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/getCurrentEditorContext',
                adapter: async (args, impl) => await impl(),
                getProgressText: args => 'Getting current editor context...',
                schema: {
                    name: "getCurrentEditorContext",
                    description: "Gets the context (file path, cursor line, selected text, and surrounding code lines) of the active document in the editor.",
                    parameters: {
                        type: "object",
                        properties: {}
                    }
                }
            },
            {
                name: 'readFile',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/readFile',
                adapter: async (args, impl) => await impl(args.filePath),
                getProgressText: args => 'Reading file ' + (args.filePath || '') + '...',
                schema: {
                    name: "readFile",
                    description: "Reads the entire content of a file in the workspace. Use this to inspect file contents before editing or analyzing. File must exist on disk.",
                    parameters: {
                        type: "object",
                        properties: {
                            filePath: {
                                type: "string",
                                description: "The relative or absolute path of the file to read. Ensure correct capitalization and slashes."
                            }
                        },
                        required: ["filePath"]
                    }
                }
            },
            {
                name: 'getContextAtLocation',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/getContextAtLocation',
                adapter: async (args, impl) => await impl(args.filePath, Math.max(0, parseInt(args.line, 10) - 1), Math.max(0, parseInt(args.character, 10) - 1), parseInt(args.callStackLevel || '2', 10)),
                getProgressText: args => 'Getting context at location ' + (args.filePath || '') + ':' + (args.line || 0) + '...',
                schema: {
                    name: "getContextAtLocation",
                    description: "Gets the code context (e.g. method/class body) enclosing the specified 1-based line and character location, and traces its callers recursively. Useful for understanding symbol context.",
                    parameters: {
                        type: "object",
                        properties: {
                            filePath: {
                                type: "string",
                                description: "The relative or absolute path of the file containing the symbol. Must exist on disk."
                            },
                            line: {
                                type: "integer",
                                description: "The 1-based line number inside the file (1 is the first line). Must be positive."
                            },
                            character: {
                                type: "integer",
                                description: "The 1-based character/column position on the line (1 is the first column). Must be positive."
                            },
                            callStackLevel: {
                                type: "integer",
                                description: "Optional. Number of call stack levels to trace recursively. Defaults to 2."
                            }
                        },
                        required: ["filePath", "line", "character"]
                    }
                }
            },
            {
                name: 'findDefinition',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/findDefinition',
                adapter: async (args, impl) => await impl(args.filePath, Math.max(0, parseInt(args.line, 10) - 1), Math.max(0, parseInt(args.character, 10) - 1)),
                getProgressText: args => 'Finding definition...',
                schema: {
                    name: "findDefinition",
                    description: "Finds the definition location (file, range) of the symbol at the given 1-based line and character location using language server providers.",
                    parameters: {
                        type: "object",
                        properties: {
                            filePath: {
                                type: "string",
                                description: "The relative or absolute path of the file containing the symbol reference. Must exist."
                            },
                            line: {
                                type: "integer",
                                description: "The 1-based line number where the reference is located (1 is the first line). Must be positive."
                            },
                            character: {
                                type: "integer",
                                description: "The 1-based character/column position of the symbol reference (1 is the first column). Must be positive."
                            }
                        },
                        required: ["filePath", "line", "character"]
                    }
                }
            },
            {
                name: 'findReferences',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/findReferences',
                adapter: async (args, impl) => await impl(args.filePath, Math.max(0, parseInt(args.line, 10) - 1), Math.max(0, parseInt(args.character, 10) - 1)),
                getProgressText: args => 'Finding references...',
                schema: {
                    name: "findReferences",
                    description: "Finds all references (locations and line texts) of the symbol at the given 1-based line and character location across the workspace.",
                    parameters: {
                        type: "object",
                        properties: {
                            filePath: {
                                type: "string",
                                description: "The relative or absolute path of the file containing the symbol reference. Must exist."
                            },
                            line: {
                                type: "integer",
                                description: "The 1-based line number where the reference is located (1 is the first line). Must be positive."
                            },
                            character: {
                                type: "integer",
                                description: "The 1-based character/column position of the symbol reference (1 is the first column). Must be positive."
                            }
                        },
                        required: ["filePath", "line", "character"]
                    }
                }
            },
            {
                name: 'getSymbolsInFile',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/getSymbolsInFile',
                adapter: async (args, impl) => await impl(args.filePath),
                getProgressText: args => 'Getting symbols in ' + (args.filePath || '') + '...',
                schema: {
                    name: "getSymbolsInFile",
                    description: "Gets structural symbols (classes, methods, variables, interface declarations, etc.) defined in a specific file. Target file must exist.",
                    parameters: {
                        type: "object",
                        properties: {
                            filePath: {
                                type: "string",
                                description: "The relative or absolute path of the file."
                            }
                        },
                        required: ["filePath"]
                    }
                }
            },
            {
                name: 'getDiagnostics',
                category: 'Coding Assistant',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/core_coding_assistant/getDiagnostics',
                adapter: async (args, impl) => await impl(args.filePath),
                getProgressText: args => 'Getting diagnostics...',
                schema: {
                    name: "getDiagnostics",
                    description: "Gets the active compiler/linter diagnostics (errors, warnings, hints, etc.) in the workspace or for a specific file.",
                    parameters: {
                        type: "object",
                        properties: {
                            filePath: {
                                type: "string",
                                description: "Optional relative or absolute path of the file to filter diagnostics. If omitted, returns diagnostics for the entire workspace."
                            }
                        }
                    }
                }
            },
            {
                name: 'searchWeb',
                category: 'Misc',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/misc/webSearch',
                adapter: async (args, impl) => await impl(args.query),
                getProgressText: args => 'Searching the web for "' + (args.query || '') + '"...',
                schema: {
                    name: "searchWeb",
                    description: "Performs a web search for a given query. Returns a summary of relevant information along with URL citations. Use this when you need up-to-date information, news, or factual answers that you don't know.",
                    parameters: {
                        type: "object",
                        properties: {
                            query: {
                                type: "string",
                                description: "The search query to look up on the internet."
                            }
                        },
                        required: ["query"]
                    }
                }
            },
            {
                name: 'getCurrentDateTime',
                category: 'Misc',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/misc/getCurrentDateTime',
                adapter: async (args, impl) => await impl(args),
                getProgressText: args => 'Getting current date and time...',
                schema: {
                    name: "getCurrentDateTime",
                    description: "Gets the current date and time. Use this to determine the current date, time, and day of the week if you are unsure.",
                    parameters: {
                        type: "object",
                        properties: {},
                        required: []
                    }
                }
            },
            {
                name: 'executeHttpRequest',
                category: 'Misc',
                isDestructive: false,
                isPatchingTool: false,
                modulePath: './tools/misc/executeHttpRequest',
                adapter: async (args, impl) => await impl(args.url, args.method, args.body),
                getProgressText: args => 'Executing HTTP request to "' + (args.url || '') + '"...',
                schema: {
                    name: "executeHttpRequest",
                    description: "Executes an arbitrary HTTP request (GET or POST) and returns the raw response body. Use this to fetch specific URLs, interact with APIs, or follow up on search results. You must parse the raw result yourself."
                        + " NOTE: please use sparringly. Unlike the search_web tool, this tool causes the context window to explode and provides a suboptimal user experience. Only execute this tool if it is strictly necessary or if the user explicitly asks you to look for something online.",
                    parameters: {
                        type: "object",
                        properties: {
                            url: { type: "string", description: "The full URL to request." },
                            method: { type: "string", description: "The HTTP method (GET or POST)." },
                            body: { type: "string", description: "The request body for POST requests (optional)." }
                        },
                        required: ["url", "method"]
                    }
                }
            },
        ];
    }

    getTools() {
        return this.tools;
    }

    getToolDefinition(name) {
        return this.tools.find(t => t.name === name);
    }
}

export default new ToolDefinitionProvider();
